flare_portal/experiments/models/modules.py

```diff
@@ -0,0 +1,139 @@
+
+from .. import constants
+
+
+class BaseModule(models.Model):
+    experiment = models.ForeignKey(
+        "experiments.Experiment", on_delete=models.CASCADE, related_name="modules"
+    )
+    sortorder = models.PositiveIntegerField(default=0)
+
+    objects = InheritanceManager()
+
+    class Meta:
+        ordering = ["sortorder"]
+
+    @classmethod
+    def get_module_camel_case(cls) -> str:
+        return cls.__name__.strip("Module")
+
+    @classmethod
+    def get_module_name(cls) -> str:
+        return camel_case_to_spaces(cls.get_module_camel_case())
+
+    @classmethod
+    def get_module_snake_case(cls) -> str:
+        return cls.get_module_name().replace(" ", "_")
+
+    @classmethod
+    def get_module_tag(cls) -> str:
+        return cls.get_module_snake_case().upper()
+
+    @classmethod
+    def get_module_slug(cls) -> str:
+        return slugify(cls.get_module_name())
+
+    @classmethod
+    def get_create_path_name(cls) -> str:
+        module_snake_case = cls.get_module_snake_case()
+        return f"{module_snake_case}_create"
+
+    @classmethod
+    def get_create_path(cls) -> str:
+        module_slug = cls.get_module_slug()
+        return (
+            "projects/<int:project_pk>/experiments/<int:experiment_pk>/modules/"
+            f"{module_slug}/add/"
+        )
+
+    @classmethod
+    def get_update_path_name(cls) -> str:
+        module_snake_case = cls.get_module_snake_case()
+        return f"{module_snake_case}_update"
+
+    @classmethod
+    def get_update_path(cls) -> str:
+        module_slug = cls.get_module_slug()
+        return (
+            "projects/<int:project_pk>/experiments/<int:experiment_pk>/modules/"
+            f"{module_slug}/<int:module_pk>/"
+        )
+
+    @classmethod
+    def get_delete_path_name(cls) -> str:
+        module_snake_case = cls.get_module_snake_case()
+        return f"{module_snake_case}_delete"
+
+    @classmethod
+    def get_delete_path(cls) -> str:
+        module_slug = cls.get_module_slug()
+        return (
+            "projects/<int:project_pk>/experiments/<int:experiment_pk>/modules/"
+            f"{module_slug}/<int:module_pk>/delete/"
+        )
+
+    def get_module_config(self) -> constants.ModuleConfigType:
+        raise NotImplementedError()
+
+    def get_module_description(self) -> str:
+        """Short description of module configuration"""
+        return ""
+
+    def __str__(self) -> str:
+        return f"PK: {self.pk} - Sort order: {self.sortorder}"
+
+
+class FearConditioningModule(BaseModule):
+    PHASES = Choices(
+        ("habituation", "Habituation"),
+        ("acquisition", "Acquisition"),
+        ("generalisation", "Generalisation"),
+        ("extinction", "Extinction"),
+        ("return_of_fear", "Return of fear"),
+    )
+    phase = models.CharField(max_length=24, choices=PHASES, default=PHASES.habituation)
+    trials_per_stimulus = models.PositiveIntegerField(default=0)
```

Comment:
So we show an equal number of trials for each stimulus in random order (with no more than 2 same stimuli in a row)?

Reply:
As far as I'm aware, there should be an equal number of trials with both the CS+ and the CS-. I don't think we can avoid having more than 2 same stimuli in a row because if we do, the CS would just swap back and forth between CS+ and CS- which I don't think is what they want.